build.assets/arch.mk

```diff
@@ -0,0 +1,11 @@
+# These variables are extracted from build.assets/Makefile so they can be imported
+# by other Makefiles
+ifeq ($(origin ARCH), undefined) # avoid use of "ARCH ?= $(...)" as the lazy loading will repeatedly re-run the same shell command
+	ARCH := $(shell go env GOARCH)
```

Comment:
This will unconditionally run `go` every time `make` is invoked and spit out errors if `go` is not installed. This currently happens and has caused confusion, so we can try to fix that now. How about this:

```
ifeq ($(origin GOARCH), undefined)
        GOARCH := $(go env GOARCH 2>/dev/null)
endif

ARCH = $(or $(GOARCH),$(error go not installed to determine ARCH))
```

This will set `GOARCH` to whatever `go` has if the `go` command exists, otherwise it will be empty. Then `$(ARCH)` is set to that using a recursive assignment, so it does not get evaluated unless expanded. if `$(GOARCH)` is empty, we get an error instead.

I realise now that what I said about the existing errors and confusion refer to the main Makefile not this build.assets/Makefile. Ultimately we should unify these I think, but for now at least this should surface errors early if the go toolchain is not installed and we try to use `go env GOARCH`.

Reply:
The risk here is we're turning a soft failure into a hard failure -- which then requires me to go follow up and fix the consumers of this Makefile that don't have go available.  While that is laudable, it is a scope creep for into fixing something that "isn't broken" (builds will succeed), in a PR that is otherwise fixing something very broken (no tag builds will succeed).

I'll take a look at this, but if it ends up flushing out a lot of failures, I'm going to pass on it for this PR.

Reply:
Updated in https://github.com/gravitational/teleport/pull/36974/commits/fb5949add35ac7c6b27c0964d9c7be0b3a26ce67.  This has passed my local testing, but we won't know for sure it is safe until after I finish a tag build and promote.

Reply:
testing shows that making this a hard fail is going to break a bunch of stuff:

https://github.com/gravitational/teleport.e/actions/runs/7618315809/job/20749749082

As such, I'm going to skip it for this PR.